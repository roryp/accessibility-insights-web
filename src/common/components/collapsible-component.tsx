// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { css } from '@uifabric/utilities';
import { ActionButton } from 'office-ui-fabric-react';
import * as React from 'react';

export interface CollapsibleComponentProps {
    header: JSX.Element;
    content: JSX.Element;
    contentClassName?: string;
    containerClassName?: string;
}

interface CollapsibleComponentState {
    showContent: boolean;
}

export class CollapsibleComponent extends React.Component<
    CollapsibleComponentProps,
    CollapsibleComponentState
> {
    private readonly iconNameDown = 'ChevronDown';
    private readonly iconNameUp = 'ChevronRight';

    constructor(props: CollapsibleComponentProps) {
        super(props);
        this.state = { showContent: true };
    }

    private onClick = (): void => {
        const newState = !this.state.showContent;
        this.setState({ showContent: newState });
    };

    public render(): JSX.Element {
        const showContent = this.state.showContent;
        let iconName = this.iconNameUp;
        let content: JSX.Element | null = null;

        if (showContent) {
            iconName = this.iconNameDown;
            content = (
                <div className={css(this.props.contentClassName, 'collapsible-content')}>
                    {this.props.content}
                </div>
            );
        }

        return (
            <div className={css(this.props.containerClassName, 'collapsible-component')}>
                <ActionButton
                    className="collapsible"
                    iconProps={{ iconName: iconName, class: 'collapsible-icon' }}
                    onClick={this.onClick}
                    aria-expanded={showContent}
                >
                    <span className="collapsible-title">{this.props.header}</span>
                </ActionButton>
                {content}
            </div>
        );
    }
}
