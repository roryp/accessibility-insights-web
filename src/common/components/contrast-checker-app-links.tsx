// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { windowsProductName } from 'content/strings/application';
import * as React from 'react';
import { NamedFC } from '../react/named-fc';
import { NewTabLink } from './new-tab-link';

export const WindowsContrastCheckerAppLink = NamedFC('WindowsContrastCheckerAppLink', () => (
    <NewTabLink href="https://go.microsoft.com/fwlink/?linkid=2075365">
        {windowsProductName}
    </NewTabLink>
));

export const MacContrastCheckerAppLink = NamedFC('MacContrastCheckerAppLink', () => (
    <NewTabLink href="https://developer.paciellogroup.com/resources/contrastanalyser/">
        Colour Contrast Analyser
    </NewTabLink>
));
