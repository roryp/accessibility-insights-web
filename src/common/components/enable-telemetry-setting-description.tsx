// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import { NamedFC } from '../react/named-fc';
import { PrivacyStatementText, PrivacyStatementTextDeps } from './privacy-statement-text';
import { TelemetryNotice, TelemetryNoticeDeps } from './telemetry-notice';

export type EnableTelemetrySettingDescriptionDeps = TelemetryNoticeDeps & PrivacyStatementTextDeps;

export type EnableTelemetrySettingDescriptionProps = {
    deps: EnableTelemetrySettingDescriptionDeps;
};

export const EnableTelemetrySettingDescription = NamedFC<EnableTelemetrySettingDescriptionProps>(
    'EnableTelemetrySettingDescription',
    props => (
        <>
            <TelemetryNotice deps={props.deps} />
            <PrivacyStatementText deps={props.deps} />
        </>
    ),
);
