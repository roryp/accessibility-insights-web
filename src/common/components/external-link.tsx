// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Link } from 'office-ui-fabric-react';
import * as React from 'react';

import { ActionInitiators } from '../action/action-initiator';
import { NamedFC } from '../react/named-fc';

export type ExternalLinkDeps = {
    actionInitiators: Pick<ActionInitiators, 'openExternalLink'>;
};

export type ExternalLinkProps = {
    deps: ExternalLinkDeps;
    href: string;
    title?: string;
};

export const ExternalLink = NamedFC<ExternalLinkProps>(
    'ExternalLink',
    ({ deps, href, title, children }) => {
        const onClick = e => deps.actionInitiators.openExternalLink(e, { href });
        return (
            <Link
                className="insights-link"
                target="_blank"
                href={href}
                title={title}
                onClick={onClick}
            >
                {children}
            </Link>
        );
    },
);
