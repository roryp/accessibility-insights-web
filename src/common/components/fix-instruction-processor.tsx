// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';
import * as styles from './cards/fix-instruction-color-box.scss';

type ColorMatch = {
    splitIndex: number;
    colorHexValue: string;
};

export class FixInstructionProcessor {
    private readonly colorValueMatcher = `(#[0-9a-f]{6})`;
    private readonly foregroundColorText = 'foreground color: ';
    private readonly foregroundRegExp = new RegExp(
        `${this.foregroundColorText}${this.colorValueMatcher}`,
        'i',
    );
    private readonly backgroundColorText = 'background color: ';
    private readonly backgroundRegExp = new RegExp(
        `${this.backgroundColorText}${this.colorValueMatcher}`,
        'i',
    );

    public process(fixInstruction: string): JSX.Element {
        const matches = this.getColorMatches(fixInstruction);
        return this.splitFixInstruction(fixInstruction, matches);
    }

    private getColorMatches(fixInstruction: string): ColorMatch[] {
        const foregroundMatch = this.getColorMatch(fixInstruction, this.foregroundRegExp);
        const backgroundMatch = this.getColorMatch(fixInstruction, this.backgroundRegExp);

        return [foregroundMatch, backgroundMatch].filter(match => match != null) as ColorMatch[];
    }

    private getColorMatch(fixInstruction: string, colorRegex: RegExp): ColorMatch | null {
        if (!colorRegex.test(fixInstruction)) {
            return null;
        }

        const match = colorRegex.exec(fixInstruction);
        if (match == null || match[1] == null) {
            return null;
        }

        const colorHexValue = match[1];

        return {
            splitIndex: match.index + this.foregroundColorText.length + colorHexValue.length,
            colorHexValue,
        };
    }

    private splitFixInstruction(fixInstruction: string, matches: ColorMatch[]): JSX.Element {
        const sortedMatches = matches.sort((a, b) => a.splitIndex - b.splitIndex);

        if (sortedMatches.length === 0) {
            return <>{fixInstruction}</>;
        }

        let insertionIndex = 0;
        let keyIndex = 0;

        const result: JSX.Element[] = [];

        sortedMatches.forEach(match => {
            const endIndex = match.splitIndex - match.colorHexValue.length;
            const substring = fixInstruction.substring(insertionIndex, endIndex);

            result.push(<span key={`instruction-split-${keyIndex++}`}>{substring}</span>);

            result.push(this.createColorBox(match.colorHexValue, keyIndex++));

            insertionIndex = endIndex;
        });

        const coda = fixInstruction.substr(insertionIndex);

        result.push(<span key={`instruction-split-${keyIndex++}`}>{coda}</span>);

        return (
            <>
                <span aria-hidden="true">{result}</span>
                <span className={styles.screenReaderOnly}>{fixInstruction}</span>
            </>
        );
    }

    private createColorBox(colorHexValue: string, keyIndex: number): JSX.Element {
        return (
            <span
                key={`instruction-split-${keyIndex}`}
                className={styles.fixInstructionColorBox}
                style={{ backgroundColor: colorHexValue }}
            />
        );
    }
}
