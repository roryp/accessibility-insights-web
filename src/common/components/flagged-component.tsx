// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

import { FeatureFlagStoreData } from '../types/store-data/feature-flag-store-data';

export interface FlaggedComponentProps {
    featureFlagStoreData: FeatureFlagStoreData;
    featureFlag: string;
    enableJSXElement: JSX.Element;
    disableJSXElement?: JSX.Element;
}

export class FlaggedComponent extends React.Component<FlaggedComponentProps> {
    public render(): JSX.Element | null {
        const flagName = this.props.featureFlag;

        if (this.props.featureFlagStoreData != null && this.props.featureFlagStoreData[flagName]) {
            return this.props.enableJSXElement;
        }

        return this.props.disableJSXElement || null;
    }
}
