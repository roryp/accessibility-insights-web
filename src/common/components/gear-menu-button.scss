// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
@import '../../common/styles/colors.scss';

:global(.ms-Button).gear-menu-button:focus::after {
    border: 1px $neutral-0 solid;
}

:global(.high-contrast-theme) {
    .gear-menu-button {
        color: $neutral-alpha-90;
    }

    .gear-menu-button-callout {
        box-shadow: $always-white 0px 0px 1px 0px !important;
        border-radius: unset;
    }
}
