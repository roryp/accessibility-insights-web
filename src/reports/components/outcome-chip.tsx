// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { NamedFC } from 'common/react/named-fc';
import * as React from 'react';

import { OutcomeIcon } from './outcome-icon';
import { OutcomeType, outcomeTypeSemantics } from './outcome-type';

export const failureCountAutomationId = 'count';

interface OutcomeChipProps {
    outcomeType: OutcomeType;
    count: number;
}

export const OutcomeChip = NamedFC<OutcomeChipProps>('OutcomeChip', props => {
    const { outcomeType, count } = props;
    const { pastTense } = outcomeTypeSemantics[outcomeType];

    const text = `${count} ${pastTense}`;

    return (
        <span className={'outcome-chip outcome-chip-' + outcomeType} title={text}>
            <span className="icon">
                <OutcomeIcon outcomeType={outcomeType} />
            </span>
            <span
                data-automation-id={failureCountAutomationId}
                className="count"
                aria-hidden="true"
            >
                {' '}
                {count}
            </span>
            <span className="screen-reader-only">{text}</span>
        </span>
    );
});
