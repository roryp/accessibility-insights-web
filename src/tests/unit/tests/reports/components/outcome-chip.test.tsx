// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { shallow } from 'enzyme';
import * as React from 'react';
import { OutcomeChip } from 'reports/components/outcome-chip';

describe('OutcomeChip', () => {
    describe('render', () => {
        test('pass', () => {
            const wrapper = shallow(<OutcomeChip outcomeType="pass" count={3} />);
            expect(wrapper.getElement()).toMatchSnapshot();
        });

        test('incomplete', () => {
            const wrapper = shallow(<OutcomeChip outcomeType="incomplete" count={2} />);
            expect(wrapper.getElement()).toMatchSnapshot();
        });

        test('fail', () => {
            const wrapper = shallow(<OutcomeChip outcomeType="fail" count={4} />);
            expect(wrapper.getElement()).toMatchSnapshot();
        });

        test('inapplicable', () => {
            const wrapper = shallow(<OutcomeChip outcomeType="inapplicable" count={4} />);
            expect(wrapper.getElement()).toMatchSnapshot();
        });
    });
});
