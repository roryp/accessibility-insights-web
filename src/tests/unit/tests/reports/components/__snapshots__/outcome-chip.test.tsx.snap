// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutcomeChip render fail 1`] = `
<span
  className="outcome-chip outcome-chip-fail"
  title="4 Failed"
>
  <span
    className="icon"
  >
    <OutcomeIcon
      outcomeType="fail"
    />
  </span>
  <span
    aria-hidden="true"
    className="count"
    data-automation-id="count"
  >
     
    4
  </span>
  <span
    className="screen-reader-only"
  >
    4 Failed
  </span>
</span>
`;

exports[`OutcomeChip render inapplicable 1`] = `
<span
  className="outcome-chip outcome-chip-inapplicable"
  title="4 Not applicable"
>
  <span
    className="icon"
  >
    <OutcomeIcon
      outcomeType="inapplicable"
    />
  </span>
  <span
    aria-hidden="true"
    className="count"
    data-automation-id="count"
  >
     
    4
  </span>
  <span
    className="screen-reader-only"
  >
    4 Not applicable
  </span>
</span>
`;

exports[`OutcomeChip render incomplete 1`] = `
<span
  className="outcome-chip outcome-chip-incomplete"
  title="2 Incomplete"
>
  <span
    className="icon"
  >
    <OutcomeIcon
      outcomeType="incomplete"
    />
  </span>
  <span
    aria-hidden="true"
    className="count"
    data-automation-id="count"
  >
     
    2
  </span>
  <span
    className="screen-reader-only"
  >
    2 Incomplete
  </span>
</span>
`;

exports[`OutcomeChip render pass 1`] = `
<span
  className="outcome-chip outcome-chip-pass"
  title="3 Passed"
>
  <span
    className="icon"
  >
    <OutcomeIcon
      outcomeType="pass"
    />
  </span>
  <span
    aria-hidden="true"
    className="count"
    data-automation-id="count"
  >
     
    3
  </span>
  <span
    className="screen-reader-only"
  >
    3 Passed
  </span>
</span>
`;
